### ISSUE TITLE
[Question] ... Strimzi with RedHat Service mesh compatibility 

### ISSUE THREAD

**Issue by Portshift-Admin** [strimzi/strimzi-kafka-operator#3399]
Hi,
I'm runing an OpenShift cluster (4.5) and I'm using Strimzi community operator (V0.18.0 - running 3 Zookeepers and 3 Kafka) and RedHat OpenShift Service Mesh operator (Istio).

When running this setup without Istio - all the pods gets into "Running" and "Ready" state.

After adding the service mesh into the setup (each of the pods get another sidecar - istio-proxy), The ZooKeepers is running and ready but the Kafka is failing to start due to connection timeout to the Zookeeper.

```
2020-07-28 09:11:23,342 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2020-07-28 09:11:23,343 INFO Socket connection established, initiating session, client: /127.0.0.1:37132, server: localhost/127.0.0.1:2181 (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2020-07-28 09:11:23,485 INFO Session: 0x0 closed (org.apache.zookeeper.ZooKeeper) [main]
2020-07-28 09:11:23,485 INFO EventThread shut down for session: 0x0 (org.apache.zookeeper.ClientCnxn) [main-EventThread]
2020-07-28 09:11:23,488 INFO [ZooKeeperClient Kafka server] Closed. (kafka.zookeeper.ZooKeeperClient) [main]
2020-07-28 09:11:23,492 ERROR Fatal error during KafkaServer startup. Prepare to shutdown (kafka.server.KafkaServer) [main]
kafka.zookeeper.ZooKeeperClientTimeoutException: Timed out waiting for connection while in state: CONNECTING
	at kafka.zookeeper.ZooKeeperClient.$anonfun$waitUntilConnected$3(ZooKeeperClient.scala:262)
	at kafka.zookeeper.ZooKeeperClient.waitUntilConnected(ZooKeeperClient.scala:258)
	at kafka.zookeeper.ZooKeeperClient.<init>(ZooKeeperClient.scala:119)
	at kafka.zk.KafkaZkClient$.apply(KafkaZkClient.scala:1863)
	at kafka.server.KafkaServer.createZkClient$1(KafkaServer.scala:378)
	at kafka.server.KafkaServer.initZkClient(KafkaServer.scala:403)
	at kafka.server.KafkaServer.startup(KafkaServer.scala:210)
	at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:44)
	at kafka.Kafka$.main(Kafka.scala:82)
	at kafka.Kafka.main(Kafka.scala)
```
I don't see any errors in the tls sidecar.

After looking into [Istio's instructions regarding ZooKeepers](https://istio.io/latest/faq/applications/#zookeeper) we needed to add an environment variable to the zookeeper config. We didn't find any way to do this via the strimzi operator so we tried to change it in the statefullset itself. 
Unfortunately it seems like the operator change this configuration back once we restart the pod. That lead us to remove the operator so it won't change the configuration back.

My question is:
1. How can we change the zookeeper config via the operator? is there any way to do that without the operator change it back?
2. Is there any a workaround/integration with Strimzi and istio-proxy?
3. Can we somehow disable the tls sidecar with the Strimzi configuration?

**Comment by scholzj** [comment]
I do not think you can run Strimzi with the Istio sidecar injected. You will need to exclude Strimzi and all its pods from Istio and just configure Istio to pass any connections from Istio applications connecting to Kafka as regualr TCP connection. AFAIK there are users having Strimzi co-exist with Istio like this.

**Comment by Portshift-Admin** [comment]
Thanks for your response
What about changing the config maps through the operator?
In the headline of the kafka-config map it specifies that the konfig map will be reconfigured by the operator but I can't find a way to modify it in the operator kafka CRD

Is there any way to controle these kind of things through the CDR's?

environment variables , config maps values, annotations, labels?

**Comment by scholzj** [comment]
Sorry ... not sure what kind of config maps or env vars do you have in mind. Could you be more specific?

**Comment by Portshift-Admin** [comment]
For example, kafka-config.service.config
![image](https://user-images.githubusercontent.com/41139662/88837165-899f0100-d1e0-11ea-8f6b-6c896cfe850a.png)

and Zookeeper config

**Comment by scholzj** [comment]
Ahh, ok. These are generated from the Kafka CR. Different parts from different sections. The only thing which you can changed directly is in `.spec.kafka.confg` - but it allows only some options. Other can be changed through for example listeners configuration, storage etc. - but these can be changed only to some extent.

Operators will be always opinionated because they need to understand how the system is configured in order to operate it. adn that is why you cannot really freely configure everything.

**Comment by Portshift-Admin** [comment]
let me rephrase.
We need to connect to the authorization mechanism of Kafka (we do it with other vendors like Confulence and Yolean)
For that we need two things:
One to add the following line to the service.config: 
```
authorizer.class.name:KafkaAuthz.PortshiftKafkaAuthz
``` 

And to add an annotation to the statefullset template of the kafka broker:
```
auth.custom/kafka: true
```

Can we do it in a way that the operator will not change it back?

**Comment by scholzj** [comment]
Adding annotation should be fairly easy. You can do it in your Kafka CR using the `template` object: https://strimzi.io/docs/operators/latest/using.html#assembly-customizing-kubernetes-resources-str

A custom authoriser is a bit more complicated. You will need to add support for it to the `api` module and some other classes. It will also have to support the `super.users` option. You can see this PR as an example - this added the OPA authorization: https://github.com/strimzi/strimzi-kafka-operator/pull/3192 ... and once you have that you can rebuild the images. 

If the authorizer is something generally useful and open source, you could also open a [proposal](https://github.com/strimzi/proposals) and it could be added directly to Strimzi.

**Comment by Portshift-Admin** [comment]
Thanks!
Before I'm closing this question.
After opening a proposal - can you please specify what is the process? I guess we will develop a solution and send an MR right?

**Comment by scholzj** [comment]
Yes. Basically in the proposal you will explain what, why and how it should be implemented. And if it is approved by the project maintainers, you can open a PR with the code, tests, docs etc.

**Comment by scholzj** [comment]
Is there still something we can help with here? Or can we close the issue? Thanks

**Comment by ReggieCarey** [comment]
I'm not entirely sure this is resolved to satisfaction. The value of a service mesh from a security standpoint is very high. Not sure why this fails to work.  Worse, I'm surprised its not documented on your website. I've spent a huge amount of time getting to this particular bug and it would have been short circuited had you mentioned incompatibility with Istio. 

1) Document the incompatibility when trying to create a cluster - some place prominent on the getting started pages at a minimum and on operatorhub.io.
1.5) Have the operator examine the metadata of the target namespace to verify service mesh is disabled
1.5a) Edit pod templates to disable sidecar injection so that it can still be deployed into sidecar enabled namespaces
2) Request that you resolve this problem so that clusters can be spun up in namespaces that have a service mesh enabled.

**Comment by scholzj** [comment]
> I'm not entirely sure this is resolved to satisfaction.

This was not resolved. It was just closed because we do not support any integration with Red Hat Service Mesh at this point.

> Worse, I'm surprised its not documented on your website.

There are many projects and tools which Strimzi does not integrate with. I'm not sure why would Red Hat Service Mesh deserve any special mentions. We normally focus on the projects we integrate with. 

I'm also not sure why would we be investing our time into detecting Red Hat Service Mesh. Normally it is the Service Mesh which tries to inject itself into a working software and breaks it. Not the other way around. So if anything, it should probably be the Service Mesh detecting if the other applications support it or not.

**Comment by ReggieCarey** [comment]
It was open source Istio service mesh that was referred to, not red hat. But I understand the position you’re taking. 

Industry is moving towards zero trust architecture and that typically involves a service mesh. It turns out that the istio solution is the most common and it’s offered out of the box by many k8s providers. (Evidenced by RedHat placing its moniker there upon)

To you last point of interoperability with service mesh, I filed a similar bug with istio. In general service meshes are supposed to be transparent to the apps that run inside them. But it sometimes requires tweaking. In this case it appears that a change in Zookeeper configuration might be a path to a solution. 

I won’t pretend to understand the complexity of the Strimzi offering. It’s a great product. But this defensive circling the wagons approach to handling bugs filed against the product is not productive. 

We all benefit from understanding the complex interoperability between the open source solutions we all rely upon. 

With greatest sincerity.

**Comment by scholzj** [comment]
I thought that Red hat Service Mesh is based on Istio so what we really talk here about is Istio integration? In any case Strimzi is an open source community. So it normally integrates with other open source communities / projects as Istio and not with vendor products. So if we ever get to something, it will be Istio integration.

And as discussed several times in the past, while the demand for Istio integration exists, it is a relatively niche thing which is not our priority. But if you or anyone else is interested in contributing and maintaining the support for Istio integration, we are more than happy to talk about it and have it part of Strimzi. I think we are very clear and transparent about this for a long time.
